src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractMultiLifetimeComponent.cs

```diff
@@ -62,54 +63,52 @@ public async Task LoadAsync()
 
         public async Task LoadCoreAsync()
         {
-            JoinableTask initializeAsyncTask;
+            await _semLock.WaitAsync();
 
-            lock (_lock)
+            if (!_instanceTaskSource.Task.IsCompleted)
             {
-                if (!_instanceTaskSource.Task.IsCompleted)
-                {
-                    (T instance, JoinableTask initializeAsyncTask) result = CreateInitializedInstance();
-                    _instanceTaskSource.SetResult(result);
-                }
+                (T instance, JoinableTask initializeAsyncTask) result = CreateInitializedInstance();
+                _instanceTaskSource.SetResult(result);
+            }
 
-                Assumes.True(_instanceTaskSource.Task.IsCompleted);
+            Assumes.True(_instanceTaskSource.Task.IsCompleted);
 
-                // Should throw TaskCanceledException if already cancelled in Dispose
-                (_, initializeAsyncTask) = _instanceTaskSource.Task.GetAwaiter().GetResult();
-            }
+            // Should throw TaskCanceledException if already cancelled in Dispose
+            (T instance, JoinableTask initializeAsyncTask)? oldInstanceTask = await _instanceTaskSource.Task;
 
-            await initializeAsyncTask;
+            _semLock.Release();
```

Comment:
The release needs to be in a `finally` block, in case an exception occurs within the critical section.